File: tasks/rezoning_geocode_stage1_parcel/code/build_census_template.R
## run this line when editing code in Rstudio
# setwd("/Users/jacobherbstman/Desktop/alderman_data/tasks/rezoning_geocode_stage1_parcel/code")
#
## interactive argument examples (mirror Makefile inputs)
# build_census_template.R --date-tag 20101201_20260212

source("../../setup_environment/code/packages.R")

parse_args <- function(args) {
  out <- list(date_tag = "20101201_20260212")
  i <- 1
  while (i <= length(args)) {
    arg <- args[[i]]
    if (startsWith(arg, "--date-tag=")) {
      out$date_tag <- sub("^--date-tag=", "", arg)
    } else if (arg == "--date-tag" && i < length(args)) {
      i <- i + 1
      out$date_tag <- args[[i]]
    }
    i <- i + 1
  }
  out
}

clean_street_address <- function(value) {
  if (is.na(value)) return("")
  stringr::str_squish(as.character(value))
}

derive_from_external <- function(address_for_external) {
  if (is.na(address_for_external)) return("")
  text <- stringr::str_squish(as.character(address_for_external))
  lowered <- tolower(text)
  bad_addr <- c(
    "",
    "nan",
    "na",
    "none",
    "nan, chicago, il",
    "na, chicago, il",
    "none, chicago, il",
    ", chicago, il",
    "chicago, il"
  )
  if (lowered %in% bad_addr) return("")
  text <- stringr::str_replace(text, ",?\\s+chicago\\s*(,\\s*)?il(?:linois)?\\s*$", "")
  text <- stringr::str_squish(stringr::str_trim(text, side = "both"))
  text <- stringr::str_replace_all(text, "^,+|,+$", "")
  stringr::str_trim(text)
}

main <- function() {
  opts <- parse_args(commandArgs(trailingOnly = TRUE))

  in_unmatched_csv <- paste0("../output/rezoning_geocode_stage1_unmatched_", opts$date_tag, ".csv")
  out_census_template_csv <- paste0("../output/census_geocoder_upload_template_", opts$date_tag, ".csv")

  unmatched <- readr::read_csv(in_unmatched_csv, show_col_types = FALSE)

  out <- unmatched %>%
    dplyr::transmute(
      external_row_id = .data$external_row_id,
      street_address = clean_street_address(.data$street_address),
      city = .data$city,
      state = .data$state,
      zip = .data$zip,
      address_for_external = .data$address_for_external
    )

  missing_street <- stringr::str_trim(out$street_address) == ""
  out$street_address[missing_street] <- vapply(
    out$address_for_external[missing_street],
    derive_from_external,
    FUN.VALUE = character(1)
  )

  out <- out %>%
    dplyr::mutate(
      city = dplyr::if_else(is.na(.data$city) | .data$city == "", "Chicago", .data$city),
      state = dplyr::if_else(is.na(.data$state) | .data$state == "", "IL", .data$state),
      zip = dplyr::if_else(is.na(.data$zip), "", .data$zip)
    ) %>%
    dplyr::select(.data$external_row_id, .data$street_address, .data$city, .data$state, .data$zip)

  dir.create("../output", recursive = TRUE, showWarnings = FALSE)
  readr::write_csv(out, out_census_template_csv, col_names = FALSE)
}

main()

File: tasks/audits/rezoning_external_geocoder_handoff/code/build_census_template.py
import argparse
import re
from pathlib import Path

import pandas as pd

BAD_ADDR = {
    "",
    "nan",
    "na",
    "none",
    "nan, chicago, il",
    "na, chicago, il",
    "none, chicago, il",
    ", chicago, il",
    "chicago, il",
}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--in-unmatched-csv", required=True)
    parser.add_argument("--out-census-template-csv", required=True)
    return parser.parse_args()


def ensure_parent(path: str) -> None:
    Path(path).parent.mkdir(parents=True, exist_ok=True)


def clean_street_address(value) -> str:
    if pd.isna(value):
        return ""
    out = str(value).strip()
    out = re.sub(r"\s+", " ", out)
    return out


def derive_from_external(address_for_external) -> str:
    if pd.isna(address_for_external):
        return ""
    text = str(address_for_external).strip()
    lowered = text.lower()
    if lowered in BAD_ADDR:
        return ""
    text = re.sub(r",?\s+chicago\s*(,\s*)?il(?:linois)?\s*$", "", text, flags=re.IGNORECASE)
    text = re.sub(r"\s+", " ", text).strip(" ,")
    return text


def main() -> int:
    args = parse_args()

    unmatched = pd.read_csv(args.in_unmatched_csv, dtype=str, low_memory=False)

    out = unmatched[["external_row_id", "street_address", "city", "state", "zip", "address_for_external"]].copy()
    out["street_address"] = out["street_address"].map(clean_street_address)

    missing = out["street_address"].str.strip().eq("")
    out.loc[missing, "street_address"] = out.loc[missing, "address_for_external"].map(derive_from_external)

    out["city"] = out["city"].fillna("Chicago")
    out["state"] = out["state"].fillna("IL")
    out["zip"] = out["zip"].fillna("")

    template = out[["external_row_id", "street_address", "city", "state", "zip"]].copy()

    ensure_parent(args.out_census_template_csv)
    template.to_csv(args.out_census_template_csv, index=False, header=False)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
